clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/controllers/AmazonNamedImageLookupController.groovy

```diff
@@ -95,11 +94,51 @@ class AmazonNamedImageLookupController {
 
     Collection<CacheData> matchesByImageId = cacheView.getAll(IMAGES.ns, imageIdentifiers)
 
-    render(matchesByName, matchesByImageId, null, lookupOptions.q, lookupOptions.region)
+    return filter(
+      render(matchesByName, matchesByImageId, lookupOptions.q, lookupOptions.region),
+      request
+    )
+  }
+
+  /**
+   * Apply tag-based filtering to the list of named images.
+   *
+   * ie. /aws/images/find?q=PackageName&tag:engine=spinnaker&tag:stage=released
+   */
+  private List<NamedImage> filter(List<NamedImage> namedImages, HttpServletRequest httpServletRequest) {
+    def tagFilters = httpServletRequest.getParameterNames().findAll {
+      it.toLowerCase().startsWith("tag:")
+    }.collectEntries { String tagParameter ->
+      [tagParameter.replaceAll("tag:", ""), httpServletRequest.getParameter(tagParameter)]
+    } as Map<String, String>
+
+    if (!tagFilters) {
+      return namedImages
+    }
+
+    return namedImages.findAll { NamedImage namedImage ->
+      def matches = tagFilters.find {
+        // find a tag that doesn't match
+        !namedImage.tags[it.key.toLowerCase()]?.equalsIgnoreCase(it.value)
+      }
+      return matches == null
+    }
   }
 
-  private List<NamedImage> render(Collection<CacheData> namedImages, Collection<CacheData> images, Collection<CacheData> imageTags, String requestedName = null, String requiredRegion = null) {
+  private List<NamedImage> render(Collection<CacheData> namedImages, Collection<CacheData> images, String requestedName = null, String requiredRegion = null) {
     Map<String, NamedImage> byImageName = [:].withDefault { new NamedImage(imageName: it) }
+
+    cacheView.getAll(IMAGES.ns, namedImages.collect {
+      (it.relationships[IMAGES.ns] ?: []).collect {
+        it
+      }
+    }.flatten() as Collection<String>).each {
+      // present tags as a merged set across all AMIs
+      byImageName[it.attributes.name as String].tags.putAll(
```

Comment:
it will prevent ability to do tag-based promotions of ami ( i.e, test -> candidate -> release ) for people that do one region and then all other 2, won't it?